File: packages/browser-domain/src/browser-hidden-analysis-loader.ts
import {
  type Infer,
  Schema,
  withParser,
  brandedNonEmptyStringSchema,
  NonEmptyStringSchema
} from '@ocentra-parent/schema-domain/effect';
import { ActivityEvidenceIdSchema, ActivityTimestampSchema } from '@ocentra-parent/schema-domain/evidence-primitives';
import {
  BrowserHiddenAnalysisDegradedReasonSchema,
  BrowserHiddenAnalysisProfileDesignIdSchema,
  BrowserHiddenAnalysisProfileDesignSchema,
  BrowserHiddenAnalysisStateSchema,
} from '@ocentra-parent/schema-domain/browser-hidden-analysis-schemas';

export const BrowserHiddenAnalysisLoaderSchemaVersion = 1;
const OptionalHiddenAnalysisLoaderTextSchema = Schema.Union(NonEmptyStringSchema, Schema.Null);

export const BrowserHiddenAnalysisLoaderRequestIdSchema = withParser(
  brandedNonEmptyStringSchema('BrowserHiddenAnalysisLoaderRequestId')
);

export const BrowserHiddenAnalysisLoaderResultIdSchema = withParser(
  brandedNonEmptyStringSchema('BrowserHiddenAnalysisLoaderResultId')
);

export const BrowserHiddenAnalysisLoaderCapabilityStateSchema = withParser(
  Schema.Literal('available', 'disabled-by-policy', 'unavailable', 'manual-required')
);

const BrowserHiddenAnalysisLoaderRequestBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(BrowserHiddenAnalysisLoaderSchemaVersion),
  loaderRequestId: BrowserHiddenAnalysisLoaderRequestIdSchema,
  requestedAt: ActivityTimestampSchema,
  profileDesign: BrowserHiddenAnalysisProfileDesignSchema,
  capabilityState: BrowserHiddenAnalysisLoaderCapabilityStateSchema,
  policyAllowsHiddenAnalysis: Schema.Boolean,
});

export const BrowserHiddenAnalysisLoaderRequestSchema = withParser(
  BrowserHiddenAnalysisLoaderRequestBaseSchema.pipe(
    Schema.filter(
      (value) =>
        browserHiddenAnalysisLoaderRequestIsConsistent(value) ||
        'Expected hidden analysis loader request to preserve policy and profile-design boundaries'
    )
  )
);

const BrowserHiddenAnalysisLoaderResultBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(BrowserHiddenAnalysisLoaderSchemaVersion),
  loaderResultId: BrowserHiddenAnalysisLoaderResultIdSchema,
  loaderRequestId: BrowserHiddenAnalysisLoaderRequestIdSchema,
  producedAt: ActivityTimestampSchema,
  profileDesignId: BrowserHiddenAnalysisProfileDesignIdSchema,
  sourceEvidenceIds: Schema.Array(ActivityEvidenceIdSchema),
  state: BrowserHiddenAnalysisStateSchema,
  loaderProofRef: OptionalHiddenAnalysisLoaderTextSchema,
  degradedReasons: Schema.Array(BrowserHiddenAnalysisDegradedReasonSchema),
  loadedByHiddenAdapter: Schema.Boolean,
  pageBodyCaptured: Schema.Boolean,
  transcriptTextCaptured: Schema.Boolean,
});

export const BrowserHiddenAnalysisLoaderResultSchema = withParser(
  BrowserHiddenAnalysisLoaderResultBaseSchema.pipe(
    Schema.filter(
      (value) =>
        browserHiddenAnalysisLoaderResultIsConsistent(value) ||
        'Expected hidden analysis loader result to preserve proof and no-capture boundaries'
    )
  )
);

export const decodeBrowserHiddenAnalysisLoaderRequest = Schema.decodeUnknownSync(
  BrowserHiddenAnalysisLoaderRequestSchema
);
export const decodeBrowserHiddenAnalysisLoaderResult = Schema.decodeUnknownSync(
  BrowserHiddenAnalysisLoaderResultSchema
);

export type BrowserHiddenAnalysisLoaderRequestId = Infer<typeof BrowserHiddenAnalysisLoaderRequestIdSchema>;
export type BrowserHiddenAnalysisLoaderResultId = Infer<typeof BrowserHiddenAnalysisLoaderResultIdSchema>;
export type BrowserHiddenAnalysisLoaderCapabilityState = Infer<typeof BrowserHiddenAnalysisLoaderCapabilityStateSchema>;
export type BrowserHiddenAnalysisLoaderRequest = Infer<typeof BrowserHiddenAnalysisLoaderRequestSchema>;
export type BrowserHiddenAnalysisLoaderResult = Infer<typeof BrowserHiddenAnalysisLoaderResultSchema>;

export function planBrowserHiddenAnalysisLoader(input: unknown): BrowserHiddenAnalysisLoaderResult {
  const request = decodeBrowserHiddenAnalysisLoaderRequest(input);
  const result = hiddenAnalysisLoaderResultForRequest(request);

  return decodeBrowserHiddenAnalysisLoaderResult(result);
}

function browserHiddenAnalysisLoaderRequestIsConsistent(
  value: Infer<typeof BrowserHiddenAnalysisLoaderRequestBaseSchema>
) {
  if (!value.policyAllowsHiddenAnalysis) {
    return value.capabilityState === 'disabled-by-policy';
  }
  if (value.capabilityState === 'available') {
    return value.profileDesign.state === 'queued' || value.profileDesign.state === 'loading';
  }
  return true;
}

function browserHiddenAnalysisLoaderResultIsConsistent(
  value: Infer<typeof BrowserHiddenAnalysisLoaderResultBaseSchema>
) {
  if (value.pageBodyCaptured || value.transcriptTextCaptured || value.sourceEvidenceIds.length === 0) {
    return false;
  }
  if (BrowserHiddenAnalysisLoadedStates.includes(value.state)) {
    return loadedHiddenAnalysisResultIsConsistent(value);
  }
  if (BrowserHiddenAnalysisPendingStates.includes(value.state)) {
    return pendingHiddenAnalysisResultIsConsistent(value);
  }
  return degradedHiddenAnalysisResultIsConsistent(value);
}

type BrowserHiddenAnalysisLoaderResultCandidate = Infer<typeof BrowserHiddenAnalysisLoaderResultBaseSchema>;

const BrowserHiddenAnalysisLoadedStates: ReadonlyArray<BrowserHiddenAnalysisLoaderResultCandidate['state']> = [
  'metadata-only',
  'analysis-ready',
] as const;

const BrowserHiddenAnalysisPendingStates: ReadonlyArray<BrowserHiddenAnalysisLoaderResultCandidate['state']> = [
  'queued',
  'loading',
] as const;

function loadedHiddenAnalysisResultIsConsistent(value: BrowserHiddenAnalysisLoaderResultCandidate): boolean {
  return value.loadedByHiddenAdapter && value.loaderProofRef !== null && value.degradedReasons.length === 0;
}

function pendingHiddenAnalysisResultIsConsistent(value: BrowserHiddenAnalysisLoaderResultCandidate): boolean {
  return !value.loadedByHiddenAdapter && value.loaderProofRef === null && value.degradedReasons.length === 0;
}

function degradedHiddenAnalysisResultIsConsistent(value: BrowserHiddenAnalysisLoaderResultCandidate): boolean {
  return !value.loadedByHiddenAdapter && value.loaderProofRef === null && value.degradedReasons.length > 0;
}

function hiddenAnalysisLoaderResultForRequest(request: BrowserHiddenAnalysisLoaderRequest) {
  if (!request.policyAllowsHiddenAnalysis) {
    return manualRequiredLoaderResult(request, ['disabled-by-policy']);
  }
  if (request.capabilityState === 'unavailable') {
    return manualRequiredLoaderResult(request, ['loader-unavailable']);
  }
  if (request.capabilityState === 'manual-required') {
    return manualRequiredLoaderResult(request, ['manual-required']);
  }

  return {
    schemaVersion: BrowserHiddenAnalysisLoaderSchemaVersion,
    loaderResultId: `${request.loaderRequestId}-result`,
    loaderRequestId: request.loaderRequestId,
    producedAt: request.requestedAt,
    profileDesignId: request.profileDesign.designId,
    sourceEvidenceIds: request.profileDesign.sourceEvidenceIds,
    state: 'loading',
    loaderProofRef: null,
    degradedReasons: [],
    loadedByHiddenAdapter: false,
    pageBodyCaptured: false,
    transcriptTextCaptured: false,
  };
}

function manualRequiredLoaderResult(
  request: BrowserHiddenAnalysisLoaderRequest,
  degradedReasons: Array<Infer<typeof BrowserHiddenAnalysisDegradedReasonSchema>>
) {
  return {
    schemaVersion: BrowserHiddenAnalysisLoaderSchemaVersion,
    loaderResultId: `${request.loaderRequestId}-result`,
    loaderRequestId: request.loaderRequestId,
    producedAt: request.requestedAt,
    profileDesignId: request.profileDesign.designId,
    sourceEvidenceIds: request.profileDesign.sourceEvidenceIds,
    state: 'manual-required',
    loaderProofRef: null,
    degradedReasons,
    loadedByHiddenAdapter: false,
    pageBodyCaptured: false,
    transcriptTextCaptured: false,
  };
}

File: crates/browser-core/src/browser_hidden_analysis_loader.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BrowserHiddenAnalysisLoaderCapabilityState {
    Available,
    DisabledByPolicy,
    ManualRequired,
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct BrowserHiddenAnalysisLoaderDecisionTemplate {
    pub state: &'static str,
    pub degraded_reasons: &'static [&'static str],
    pub loaded_by_hidden_adapter: bool,
    pub loader_proof_ref: Option<&'static str>,
    pub page_body_captured: bool,
    pub transcript_text_captured: bool,
}

const LOADING_TEMPLATE: BrowserHiddenAnalysisLoaderDecisionTemplate =
    BrowserHiddenAnalysisLoaderDecisionTemplate {
        state: "loading",
        degraded_reasons: &[],
        loaded_by_hidden_adapter: false,
        loader_proof_ref: None,
        page_body_captured: false,
        transcript_text_captured: false,
    };

const DISABLED_BY_POLICY_TEMPLATE: BrowserHiddenAnalysisLoaderDecisionTemplate =
    BrowserHiddenAnalysisLoaderDecisionTemplate {
        state: "manual-required",
        degraded_reasons: &["disabled-by-policy"],
        loaded_by_hidden_adapter: false,
        loader_proof_ref: None,
        page_body_captured: false,
        transcript_text_captured: false,
    };

const MANUAL_REQUIRED_TEMPLATE: BrowserHiddenAnalysisLoaderDecisionTemplate =
    BrowserHiddenAnalysisLoaderDecisionTemplate {
        state: "manual-required",
        degraded_reasons: &["manual-required"],
        loaded_by_hidden_adapter: false,
        loader_proof_ref: None,
        page_body_captured: false,
        transcript_text_captured: false,
    };

const UNAVAILABLE_TEMPLATE: BrowserHiddenAnalysisLoaderDecisionTemplate =
    BrowserHiddenAnalysisLoaderDecisionTemplate {
        state: "manual-required",
        degraded_reasons: &["loader-unavailable"],
        loaded_by_hidden_adapter: false,
        loader_proof_ref: None,
        page_body_captured: false,
        transcript_text_captured: false,
    };

pub fn evaluate_browser_hidden_analysis_loader(
    policy_allows_hidden_analysis: bool,
    capability_state: BrowserHiddenAnalysisLoaderCapabilityState,
) -> BrowserHiddenAnalysisLoaderDecisionTemplate {
    if !policy_allows_hidden_analysis {
        return DISABLED_BY_POLICY_TEMPLATE;
    }

    match capability_state {
        BrowserHiddenAnalysisLoaderCapabilityState::Available => LOADING_TEMPLATE,
        BrowserHiddenAnalysisLoaderCapabilityState::DisabledByPolicy => DISABLED_BY_POLICY_TEMPLATE,
        BrowserHiddenAnalysisLoaderCapabilityState::ManualRequired => MANUAL_REQUIRED_TEMPLATE,
        BrowserHiddenAnalysisLoaderCapabilityState::Unavailable => UNAVAILABLE_TEMPLATE,
    }
}

pub fn browser_hidden_analysis_loader_typescript() -> String {
    [
        "/* generated from crates/browser-core/src/browser_hidden_analysis_loader.rs */",
        "",
        "export const BrowserHiddenAnalysisLoaderDecisionTemplates = {",
        "  available: {",
        "    state: 'loading',",
        "    degradedReasons: [],",
        "    loadedByHiddenAdapter: false,",
        "    loaderProofRef: null,",
        "    pageBodyCaptured: false,",
        "    transcriptTextCaptured: false,",
        "  },",
        "  'disabled-by-policy': {",
        "    state: 'manual-required',",
        "    degradedReasons: ['disabled-by-policy'],",
        "    loadedByHiddenAdapter: false,",
        "    loaderProofRef: null,",
        "    pageBodyCaptured: false,",
        "    transcriptTextCaptured: false,",
        "  },",
        "  'manual-required': {",
        "    state: 'manual-required',",
        "    degradedReasons: ['manual-required'],",
        "    loadedByHiddenAdapter: false,",
        "    loaderProofRef: null,",
        "    pageBodyCaptured: false,",
        "    transcriptTextCaptured: false,",
        "  },",
        "  unavailable: {",
        "    state: 'manual-required',",
        "    degradedReasons: ['loader-unavailable'],",
        "    loadedByHiddenAdapter: false,",
        "    loaderProofRef: null,",
        "    pageBodyCaptured: false,",
        "    transcriptTextCaptured: false,",
        "  },",
        "} as const;",
        "",
        "export type BrowserHiddenAnalysisLoaderCapabilityState =",
        "  | 'available'",
        "  | 'disabled-by-policy'",
        "  | 'manual-required'",
        "  | 'unavailable';",
        "",
        "export type BrowserHiddenAnalysisLoaderDecisionTemplate =",
        "  (typeof BrowserHiddenAnalysisLoaderDecisionTemplates)[keyof typeof BrowserHiddenAnalysisLoaderDecisionTemplates];",
        "",
        "export function browserHiddenAnalysisLoaderDecisionTemplate(",
        "  policyAllowsHiddenAnalysis: boolean,",
        "  capabilityState: BrowserHiddenAnalysisLoaderCapabilityState",
        "): BrowserHiddenAnalysisLoaderDecisionTemplate {",
        "  if (!policyAllowsHiddenAnalysis) {",
        "    return BrowserHiddenAnalysisLoaderDecisionTemplates['disabled-by-policy'];",
        "  }",
        "",
        "  return BrowserHiddenAnalysisLoaderDecisionTemplates[capabilityState];",
        "}",
        "",
    ]
    .join("\n")
}
